src/plugins/odyssey/instance.js

```diff
@@ -0,0 +1,226 @@
+export type InstanceJSON = Compatible<{|
+  +graphJSON: GraphJSON,
+  +sortedDescriptions: $ReadOnlyArray<string>,
+  +count: number,
+|}>;
+
+/**
+ * This is the data model for a particular instance in the Odyssey Plugin.
+ * The OdysseyInstance allows adding "Entities", which are basically nodes
+ * in the Odyssey graph augmented with a type identifier, and a description.
+ * Currently, the types are restricted to types hard-coded in the
+ * [declaration file](./declaration.js) but we intend to allow instance-specified
+ * types in the future.
+ *
+ * The OdysseyInstance maintains an internal graph which actually stores the
+ * node identities, as well as added nodes. You can get a copy of this graph
+ * by calling the `.graph()` method.
+ *
+ * Entities are identified by an incrementing id (`._count`). This is convenient
+ * for implementation, although it will make reconciling simultaneous edits challenging.
+ * Once that becomes a real issue, we should consider switching to a different node/edge
+ * identification strategy.
+ */
+export class OdysseyInstance {
+  _graph: Graph;
+  _descriptions: Map<NodeAddressT, string>;
+  _count: number;
+
+  /**
+   * Construct an Odyssey Instance.
+   */
+  constructor() {
+    this._graph = new Graph();
+    this._count = 0;
+    this._descriptions = new Map();
+  }
+
+  /**
+   * Add a new node to the instance (and a corresponding node to the graph).
+   *
+   * Requires a valid node type identifier (aka a string that uniquely identifies
+   * an Odyssey node type).
+   */
+  addNode(
+    typeIdentifier: OdysseyNodeTypeIdentifier,
+    description: string
+  ): Node {
+    if (!isOdysseyNodeTypeIdentifier(typeIdentifier)) {
+      throw new Error(
+        `Tried to add node with invalid type identifier: ${typeIdentifier}`
+      );
+    }
+    const address = NodeAddress.append(
+      NODE_PREFIX,
+      typeIdentifier,
+      String(this._count)
+    );
+    this._graph.addNode(address);
+    this._count++;
+    this._descriptions.set(address, description);
+    return NullUtil.get(this.node(address));
+  }
+
+  /**
+   * Retrieve the Node corresponding to a given node address, if it exists.
+   */
+  node(address: NodeAddressT): ?Node {
+    if (!this._graph.hasNode(address)) {
+      return null;
+    }
+    const parts = NodeAddress.toParts(address);
+    // We know it is an OdysseyNodeTypeIdentifier because the instance's internal
+    // graph only has Odyssey nodes in it.
+    const nodeTypeIdentifier: OdysseyNodeTypeIdentifier = (parts[2]: any);
+    const description = NullUtil.get(this._descriptions.get(address));
+    return {address, nodeTypeIdentifier, description};
+  }
+
+  nodes(typeIdentifier?: OdysseyNodeTypeIdentifier): Iterator<Node> {
+    if (
+      typeIdentifier != null &&
+      !isOdysseyNodeTypeIdentifier(typeIdentifier)
+    ) {
+      throw new Error(`Invalid type identifier: ${typeIdentifier}`);
```

Comment:
I think it should allow bad identifiers and just not return any nodes. Safer if e.g. in the future we actually do have instances that have defined non-default types (and potentially stopped using default types)